BLE_GattServer_FOTAService/scripts/test_fota.py

```diff
@@ -0,0 +1,268 @@
+        await self.client.write_gatt_char(UUID_CONTROL_CHAR, FOTA_OP_CODE_START, True)
+
+        # Wait for the client to write XON to the status characteristic
+        count = 0
+        while count < 5:
+            try:
+                await asyncio.wait_for(self.handler.wait_for_status_notification(), timeout=10.0)
+                status = bytearray([self.handler.status_value[0]])
+                if status == FOTA_STATUS_XON:
+                    logger.info(f'Received status XON')
+                    break
+                elif status == FOTA_STATUS_XOFF:
+                    logger.info(f'Received status XOFF')
+                else:
+                    logger.warning(f'Received unknown status: {status}')
+            except asyncio.TimeoutError:
+                logger.info(f'Waiting for status notification: ({count}/{5})')
+                count += 1
+                if count >= 5:
+                    raise asyncio.TimeoutError
+
+    async def transfer_binary(self, path: str):
+        try:
+            file = open(path, 'rb')
+        except IOError:
+            logger.error(f'Could not open file {path}')
+            raise
+        data = file.read()
+
+        binary_sent = False
+        flow_paused = False
+
+        start_time = time.time()
+        while not binary_sent:
+            # Check status notifications
+            if self.handler.status_event.is_set():
+                status = bytearray([self.handler.status_value[0]])
+                if status == FOTA_STATUS_OK:
+                    logger.info(f'Received status OK')
+                elif status == FOTA_STATUS_XOFF:
+                    fragment_id = self.handler.status_value[1]
+                    logger.info(f'Received status XOFF. '
+                                f'Fragment ID = {fragment_id}')
+                    flow_paused = True
+                    self.update_fragment_id(fragment_id)
+                elif status == FOTA_STATUS_XON:
+                    fragment_id = self.handler.status_value[1]
+                    logger.info(f'Received status XON. '
+                                f'Fragment ID = {fragment_id}')
+                    flow_paused = False
+                    self.update_fragment_id(fragment_id)
+                elif status == FOTA_STATUS_SYNC_LOST:
+                    fragment_id = self.handler.status_value[1]
+                    logger.info(f'Received status SYNC LOST. '
+                                f'Fragment ID = {fragment_id}')
+                    self.update_fragment_id(fragment_id)
+                self.handler.status_event.clear()
+
+            if flow_paused:
+                continue
+
+            # Send next packet
+            n = MAXIMUM_FRAGMENT_ID * self.rollover_counter + self.fragment_id
+            packet_data = bytearray(get_chunk_n(data, FRAGMENT_SIZE, n))
+            packet_size = len(packet_data)
+            if packet_size < FRAGMENT_SIZE:
+                binary_sent = True
```

Comment:
How do you handle the case where the last fragment is not received ? It seems to go out of this loop but a new status can be posted by the server.